relayer-engine/src/storage/storage.ts

```diff
@@ -1,133 +1,235 @@
+            await this.requeueWorkflow(workflow);
+          }
         }
-      }
+      });
     } catch (e) {
       this.logger.error(
-        "Encountered an error while demoting in progress items at startup."
+        "Encountered an error while demoting in progress items at startup.",
       );
       this.logger.error(e);
     }
   }
-  getPluginStorage(plugin: Plugin): DefaultPluginStorage {
-    return new DefaultPluginStorage(this.store, plugin);
+
+  getStagingAreaKeyLock(pluginName: string): StagingAreaKeyLock {
+    return new DefaultStagingAreaKeyLock(this.store, this.logger, pluginName);
   }
 }
 
-class DefaultPluginStorage implements PluginStorage {
-  private readonly logger;
-  constructor(private readonly store: Store, readonly plugin: Plugin) {
-    this.logger = getScopedLogger(
-      [`RedisPluginStorage ${plugin.pluginName}`],
-      getLogger()
-    );
+function workflowKey(workflow: { id: string; pluginName: string }): string {
+  return `${workflow.pluginName}/${workflow.id}`;
+}
+
+class DefaultStagingAreaKeyLock implements StagingAreaKeyLock {
+  private readonly stagingAreaKey: string;
+  constructor(
+    private readonly store: RedisWrapper,
+    readonly logger: Logger,
+    pluginName: string,
+  ) {
+    this.stagingAreaKey = `${STAGING_AREA_KEY}/${sanitize(pluginName)}`;
   }
-  async addWorkflow(data: Object): Promise<void> {
-    const id_from_store = (await this.store
-      .kv()
-      .get(WORKFLOW_ID_COUNTER_KEY)) as number;
-    const workflow: Workflow = {
-      data,
-      id: id_from_store ? id_from_store : 0,
-      pluginName: this.plugin.pluginName,
-    };
-    await this.store
-      .kv()
-      .compareAndSwap(WORKFLOW_ID_COUNTER_KEY, workflow.id, workflow.id + 1);
-    await this.store.queue<Workflow>().push(workflow);
+
+  getKeys(keys: string[]): Promise<Record<string, any>> {
+    return this.store.withRedis(async redis =>
+      this.getKeysInternal(redis, keys),
+    );
   }
 
-  async getStagingArea(this: DefaultPluginStorage): Promise<Object> {
-    const key = stagingAreaKey(this.plugin);
-    const stagingArea = await this.store
-      .kv<StagingArea>(STAGING_AREA_KEY)
-      .get(key);
-    if (!stagingArea) {
-      this.logger.warn(
-        `Missing staging area for plugin ${this.plugin.pluginName}. Returning empty object`
-      );
-      return {};
-    }
-    return stagingArea;
+  private getKeysInternal(
+    redis: IRedis,
+    keys: string[],
+  ): Promise<Record<string, any>> {
+    return Promise.all(
+      keys.map(async k => {
+        const val = await redis.get(`${this.stagingAreaKey}/${k}`);
+        return [k, val !== null ? JSON.parse(val) : undefined];
+      }),
+    ).then(Object.fromEntries);
   }
 
-  saveStagingArea(
-    this: DefaultPluginStorage,
-    newStagingArea: Object
-  ): Promise<void> {
-    return this.store
-      .kv<StagingArea>(STAGING_AREA_KEY)
-      .set(stagingAreaKey(this.plugin), newStagingArea);
+  async withKey<T>(
```

Comment:
Supporting multiple listeners reading and writing the same monolithic staging area seems doomed to frequent failures so created this "keyed" api so the plugin author can intelligently pick keys to reduce contention. If the listeners do contend, the whole script can only write the result to redis if the value hasn't changed since reading it. 

In the future this should probably retry if writing fails

Reply:
It would be simpler to leave out the key mechanism but keep the transactionality. Interested in people's thoughts on what api to expose. In practice each listener is going to be a clone of the others, so maybe it doesn't matter if they race?